papaya-core/src/main/java/pl/edu/agh/papaya/service/jira/JiraOAuthClient.java

```diff
@@ -0,0 +1,131 @@
+package pl.edu.agh.papaya.service.jira;
+
+import com.google.api.client.auth.oauth.OAuthAuthorizeTemporaryTokenUrl;
+import com.google.api.client.auth.oauth.OAuthCredentialsResponse;
+import com.google.api.client.auth.oauth.OAuthParameters;
+import com.google.api.client.http.GenericUrl;
+import com.google.api.client.http.HttpRequest;
+import com.google.api.client.http.HttpRequestFactory;
+import com.google.api.client.http.HttpResponse;
+import com.google.api.client.http.javanet.NetHttpTransport;
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.security.NoSuchAlgorithmException;
+import java.security.spec.InvalidKeySpecException;
+import java.util.Optional;
+import java.util.Scanner;
+import lombok.extern.slf4j.Slf4j;
+import org.json.JSONArray;
+
+@Slf4j
+@SuppressWarnings({"ClassFanOutComplexity", "PMD.ClassNamingConventions", "PMD.BeanMembersShouldSerialize"})
+final class JiraOAuthClient {
+
+    private static final String CONSUMER_KEY = "OauthKey";
+
+    private final JiraOAuthTokenFactory oAuthTokenFactory;
+    private final URL jiraHome;
+
+    private JiraOAuthClient(URL jiraHome) throws MalformedURLException {
+        this.jiraHome = jiraHome;
+        this.oAuthTokenFactory = new JiraOAuthTokenFactory(jiraHome);
+    }
+
+    Optional<String> getAuthorizedTemporaryToken(String privateKey) {
+        try {
+            JiraOAuthGetTemporaryToken temporaryToken =
+                    oAuthTokenFactory.getTemporaryToken(CONSUMER_KEY, privateKey);
+            OAuthCredentialsResponse response = temporaryToken.execute();
+
+            return Optional.of(response.token);
+        } catch (NoSuchAlgorithmException | InvalidKeySpecException | IOException e) {
+            log.error("getAuthorizedTemporaryToken failed", e);
+            return Optional.empty();
+        }
+    }
+
+    Optional<String> getVerificationCodeUrl(String tmpToken) {
+        try {
+            URL authorizationUrl = new URL(jiraHome, "/plugins/servlet/oauth/authorize");
```

Comment:
Javadocs:
> If the spec's path component begins with a slash character "/" then the path is treated as absolute and the spec path replaces the context path.

chcemy tak?

Reply:
![image](https://user-images.githubusercontent.com/15652446/58552641-34014a00-8213-11e9-914b-acb9533859f9.png)

Czyli z tego wynikałoby, że `relativeUrl` będzie teraz `/a-guide-to-java-sockets`. Źle rozumiem?

Reply:
Trochę nie rozumiem jak się ma ten screenshot do tego co napisałem :thinking: 

Jeśli dobrze rozumiem dokumentację to:
```java
new URL(new URL("http://example.com/resource1"), "resource2"); // = "http://example.com/resource1/resource2"
new URL(new URL("http://example.com/resource1"), "/resource2"); // = "http://example.com/resource2"
```

i pytam dla pewności czy na pewno chcemy takie zachowanie.

Reply:
Hmmm, ciekawe. Testowałem i nadpisuje nawet w takim przypadku jak masz wyżej. Według RFC2396 common base URL musi się kończyć `/`. Czyli w naszym przypadku tak długo jak na końcu `jiraHome` jest `/` to za nim dopisuje ścieżkę, natomiast kiedy go nie ma to konstruktor nowego URI przycina `jiraHome` od tyłu do znalezienia slasha i po nim dokleja ścieżkę. Czyli
`http://mypage/jira + /random/endpoint = http://mypage/random/endpoint
http://mypage/jira/ + /random/endpoint = http://mypage/jira/random/endpoint`

Myślę, że niestety tak jest, że ktoś na świecie ma jire z takim url, więc dodałem handler na ten przypadek. Minus jest taki, że teraz tworzący projekt musi nam powiedzieć dokładnie pod jakim adresem jest jego jira, bo nie możemy zakładać, że jakiś randomowy path w linku możemy przyciąć, ale tak jest chyba lepiej

Reply:
![image](https://user-images.githubusercontent.com/1507072/58588484-08f11780-8260-11e9-9d58-4fdebb80b879.png)
Dokumentacja nie kłamie

Reply:
Poprawiono posługiwanie się obiektami URL. `jiraHome` kończy się teraz `/`. a wszystkie doklejane pathy go na początku nie posiadają.
Tak, nie kłamie. Tu leży moja pomyłka